File: jupyter2/ps/parametric/basics/jl/basics.jl
### begin my answer

### begin hidden
f(x) = x

g(x) = [x]

mutable struct cell
    x
end

function get(c::cell)
    c.x
end

function put!(c::cell, y)
    c.x = y
end

function h(a)
    if length(a) == 0
        nothing
    else
        a[1]
    end
end
### end hidden
### end my answer

function main()
    @assert(f(1) == 1)
    @assert(f("hello") == "hello")
    @assert(g(2) == [2])
    @assert(g("world") == ["world"])
    c0 = cell(0)
    @assert(get(c0) == 0)
    put!(c0, 42)
    c1 = cell("mimura")
    @assert(get(c1) == "mimura");
    put!(c1, "ohtake");
    @assert(get(c1) == "ohtake");
    @assert(h([]) == nothing)
    @assert(h([3]) == 3)
    @assert(h(["bye"]) == "bye")
    println("OK")
end

main()

File: jupyter2/ps/parametric/basics/rs/basics.rs
/** begin my answer */

/** begin hidden */
fn f<T>(x : T) -> T {
    x
}

fn g<T>(x : T) -> Vec<T> {
    vec![x]
}

fn h<T>(a : &[T]) -> Option<&T> {
    if a.len() == 0 {
	None
    } else {
	Some(&a[0])
    }
}

struct Cell<T> {
    x : T
}

fn get<T>(c : &Cell<T>) -> &T {
    return &c.x
}

fn put<T>(c : &mut Cell<T>, y : T) {
    c.x = y
}
/** end hidden */
/** end my answer */

fn main() {
    assert!(f(1) == 1);
    assert!(f("hello") == "hello");
    assert!(g(2) == [2]);
    assert!(g("world") == ["world"]);
    let mut c0 = Cell{x: 0};
    assert!(get(&c0) == &0);
    put(&mut c0, 42);
    let mut c1 = Cell{x: "mimura"};
    assert!(get(&c1) == &"mimura");
    put(&mut c1, "ohtake");
    assert!(get(&c1) == &"ohtake");
    assert!(h::<i64>(&vec![]) == None);
    assert!(h(&vec![3]) == Some(&3));
    assert!(h(&vec!["bye"]) == Some(&"bye"));
    println!("OK")
}
